File: libs/core/src/private/lexer/token_stream.cpp
/**
 * @file token_stream.cpp
 * @author Francesco Corso
 * @date 6/26/2026
 * @brief
 */
module;

#include <cassert>

module prism.core.token_stream;

namespace prism
{
    Token TokenStream::previous() const
    {
        return consumed_tokens_.back();
    }

    Token TokenStream::peek()
    {
        if (pending_tokens_.empty())
        {
            buffer_tokens();
        }

        return pending_tokens_.front();
    }

    Token TokenStream::consume()
    {
        advance();
        return previous();
    }

    void TokenStream::advance()
    {
        const auto token = peek();
        pending_tokens_.pop_front();
        consumed_tokens_.push_back(token);
    }

    bool TokenStream::replace_next(const Token &token)
    {
        [[maybe_unused]] const auto [kind, flags, range] = peek();
        assert(kind != TokenKind::eof);

        pending_tokens_.front() = token;
        return true;
    }

    void TokenStream::buffer_tokens()
    {
        static constexpr auto max_tokens = 1024;

        for (std::size_t i = 0; i < max_tokens; ++i)
        {
            auto token = lexer_.next();
            pending_tokens_.push_back(std::move(token));

            if (token.kind == TokenKind::eof)
            {
                break;
            }
        }
    }
} // namespace prism

File: src/Prism.Core/Parse/TokenStream.cs
// @file TokenStream.cs
//
// @copyright Copyright (c) 2026 Retro & Chill. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.

using System.Diagnostics;
using Nito.Collections;
using Prism.Core.Syntax;

namespace Prism.Core.Parse;

public sealed class TokenStream(SourceFile sourceFile)
{
    private readonly string _text = sourceFile.Text;
    private readonly Lexer _lexer = new(sourceFile);
    private readonly List<SyntaxToken> _consumedTokens = [];
    private readonly Deque<SyntaxToken> _pendingTokens = [];

    public bool AtEnd => Peek().Kind == SyntaxKind.EndOfFileToken;

    public IReadOnlyList<SyntaxToken> ConsumedTokens => _consumedTokens;

    public SyntaxToken Previous => _consumedTokens[^1];

    public SyntaxToken Peek(int count = 1)
    {
        if (_pendingTokens.Count == 0)
        {
            BufferTokens();
        }

        return _pendingTokens.Count >= count ? _pendingTokens[count - 1] : Previous;
    }

    public void Advance()
    {
        var token = Peek();
        _consumedTokens.Add(token);
        _pendingTokens.RemoveFromFront();
    }

    public bool ReplaceNext(SyntaxToken token)
    {
        var next = Peek();
        Debug.Assert(next.Kind != SyntaxKind.EndOfFileToken);

        _pendingTokens.RemoveFromFront();
        _pendingTokens.AddToFront(token);
        return true;
    }

    public bool ReplaceNext(params ReadOnlySpan<SyntaxToken> tokens)
    {
        var next = Peek();
        Debug.Assert(next.Kind != SyntaxKind.EndOfFileToken);

        _pendingTokens.RemoveFromFront();
        foreach (var token in tokens)
        {
            _pendingTokens.AddToFront(token);
        }

        return true;
    }

    private void BufferTokens()
    {
        const int maxTokens = 1024;

        for (var i = 0; i < maxTokens; i++)
        {
            var token = _lexer.Next();
            _pendingTokens.AddToBack(token);

            if (token.Kind == SyntaxKind.EndOfFileToken)
                break;
        }
    }
}
